modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/MergeJoinConverterRule.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.query.calcite.rule;
+
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelOptPlanner;
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelTraitSet;
+import org.apache.calcite.rel.PhysicalNode;
+import org.apache.calcite.rel.RelCollations;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.JoinInfo;
+import org.apache.calcite.rel.logical.LogicalJoin;
+import org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.ignite.internal.processors.query.calcite.rel.IgniteConvention;
+import org.apache.ignite.internal.processors.query.calcite.rel.IgniteMergeJoin;
+import org.apache.ignite.internal.util.typedef.F;
+
+/**
+ * Ignite Join converter.
+ */
+public class MergeJoinConverterRule extends AbstractIgniteConverterRule<LogicalJoin> {
+    /** */
+    public static final RelOptRule INSTANCE = new MergeJoinConverterRule();
+
+    /**
+     * Creates a converter.
+     */
+    public MergeJoinConverterRule() {
+        super(LogicalJoin.class, "MergeJoinConverterRule");
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean matches(RelOptRuleCall call) {
+        LogicalJoin logicalJoin = call.rel(0);
+
+        return !F.isEmpty(logicalJoin.analyzeCondition().pairs());
+    }
+
+    /** {@inheritDoc} */
+    @Override protected PhysicalNode convert(RelOptPlanner planner, RelMetadataQuery mq, LogicalJoin rel) {
+        RelOptCluster cluster = rel.getCluster();
+
+        JoinInfo joinInfo = JoinInfo.of(rel.getLeft(), rel.getRight(), rel.getCondition());
+
+        RelTraitSet leftInTraits = cluster.traitSetOf(IgniteConvention.INSTANCE)
+            .replace(RelCollations.of(joinInfo.leftKeys));
+        RelTraitSet outTraits = cluster.traitSetOf(IgniteConvention.INSTANCE)
+            .replace(RelCollations.of(joinInfo.leftKeys)); // preserve collation of the left input
```

Comment:
why do we take only left into account here ? If other cases will be produced by pass through or derive - we need additional tests here.

Reply:
because we can't preserve collation of both input's when converting from logical to physical node. There is a tests that verifies collation from ORDER BY preserved even it contains column from both table. Please see https://github.com/apache/ignite/pull/8456/files#diff-9b2afd77c4bd5e6a82744cac80418daa1804931da4828fb08d3b9c842a71ecd0R2637